packages/api/lambdas/process-s3-dead-letter-archive.js

```diff
@@ -0,0 +1,82 @@
+'use strict';
+
+const log = require('@cumulus/common/log');
+
+const { s3 } = require('@cumulus/aws-client/services');
+const { getJsonS3Object } = require('@cumulus/aws-client/S3');
+const { getKnexClient } = require('@cumulus/db');
+
+const { writeRecords } = require('./sf-event-sqs-to-db-records');
+
+async function processDeadLetterArchive({
+  knex,
+  bucket = process.env.system_bucket,
+  path = `${process.env.stackName}/dead-letter-archive/sqs/`,
+  writeRecordsFunction = writeRecords,
+  batchSize = 1000,
+}) {
+  let listObjectsResponse;
+  let continuationToken;
+  /* eslint-disable no-await-in-loop */
+  do {
+    listObjectsResponse = await s3().listObjectsV2({
+      Bucket: bucket,
+      Prefix: path,
+      ContinuationToken: continuationToken,
+      MaxKeys: batchSize,
+    }).promise();
+    continuationToken = listObjectsResponse.NextContinuationToken;
+    const deadLetterObjects = listObjectsResponse.Contents;
+    const promises = await Promise.allSettled(deadLetterObjects.map(
+      async (deadLetterObject) => {
+        const cumulusMessage = await getJsonS3Object(bucket, deadLetterObject.Key);
+        try {
+          await writeRecordsFunction({ cumulusMessage, knex });
+          return deadLetterObject.Key;
+        } catch (error) {
+          log.error(`Failed to write records from cumulusMessage for dead letter ${deadLetterObject.Key} due to '${error}'`);
+          throw error;
+        }
+      }
+    ));
+    const keysToDelete = promises.filter(
+      (prom) => prom.status === 'fulfilled'
+    ).map((prom) => ({ Key: prom.value }));
+    if (keysToDelete.length > 0) {
+      await s3().deleteObjects({
+        Bucket: bucket,
+        Delete: {
+          Objects: keysToDelete,
+        },
+      }).promise();
+    }
+  } while (listObjectsResponse.IsTruncated);
+  /* eslint-enable no-await-in-loop */
+}
+
+/**
+ * Lambda handler for AsyncOperation purposes
+ *
+ * @param {Object} event - Input payload object
+ * @param {string} [event.bucket] - Bucket containing dead letter archive (default to system bucket)
+ * @param {string} [event.key] - Dead letter archive path key
+ * @returns {Promise<undefined>}
```

Comment:
is there a difference between `Promise<undefined>` and `Promised<void>`? 🤔

Reply:
I guess `<void>` is the TS version, and `<undefined>` is the JS/ES version?

Reply:
It goes a [bit deeper than that](https://stackoverflow.com/questions/58885485/why-does-typescript-have-both-void-and-undefined) within TS, but I'm pretty sure `Promise<void>` isn't really appropriate in an EcmaScript file, as `void` there appears to be an operator with a distinct meaning.

Reply:
Then again, [ESLint](https://eslint.org/docs/rules/valid-jsdoc) appears to have the opinion that `@returns {void} constitutes valid JSDoc, so 🤷. We haven't done it anywhere else in JS files, but I don't mind starting now.

Reply:
7afb74661016d2cdd28a82ac61d8b174107a41ad